AirCasting/ABConnector/SDCardAirBeamServices.swift

```diff
@@ -0,0 +1,113 @@
+    private var currentSessionTypeExpected: Int = 0
+    
+    private var dataCharacteristicObserver: AnyHashable?
+    private var metadataCharacteristicObserver: AnyHashable?
+    
+    init(bluetoothManager: BluetoothManager) {
+        self.bluetoothManager = bluetoothManager
+    }
+    
+    func downloadData(from peripheral: CBPeripheral, progress: @escaping (SDCardData) -> Void, completion: @escaping (Result<Void, Error>) -> Void) {
+        metadataCharacteristicObserver = bluetoothManager.subscribeToCharacteristic(DOWNLOAD_META_DATA_FROM_SD_CARD_CHARACTERISTIC_UUID) { result in
+            switch result {
+            case .success(let data):
+                guard let data = data, let payload = String(data: data, encoding: .utf8) else {
+                    return
+                }
+                self.currentSessionType = self.currentSessionType.next
+                Log.info(payload)
+                if payload == "SD_SYNC_FINISH" {
+                    self.finishSync { completion(.success(())) }
+                    Log.info("Sync finished.")
+                    return
+                }
+                
+                // This will be needed when we will want to show progress in the view
+                // Payload format is ` Some string: ${number_of_entries_expected} `
+                guard let measurementsCountSting = payload.split(separator: ":").last?.trimmingCharacters(in: .whitespaces),
+                      let measurementsCount = Int(measurementsCountSting) else {
+                    Log.warning("Unexpected metadata format: (\(payload))")
+                    return
+                }
+                self.currentSessionTypeExpected = measurementsCount
+                progress(.metadata(SDCardMetaData(sessionType: self.currentSessionType!, measurementsCount: measurementsCount)))
+            case .failure(let error):
+                Log.warning("Error while receiving metadata from SD card: \(error.localizedDescription)")
+                self.finishSync { completion(.failure(error)) }
+            }
+        }
+        
+        dataCharacteristicObserver = bluetoothManager.subscribeToCharacteristic(DOWNLOAD_FROM_SD_CARD_CHARACTERISTIC_UUID) { result in
+            switch result {
+            case .success(let data):
+                guard let data = data, let payload = String(data: data, encoding: .utf8) else { return }
+                guard let sessionType = self.currentSessionType else {
+                    Log.error("Received data before first metadata payload!")
+                    return
+                }
+                progress(.chunk(SDCardDataChunk(payload: payload, sessionType: sessionType)))
+                
+
+            case .failure(let error):
+                Log.warning("Error while receiving data from SD card: \(error.localizedDescription)")
+                completion(.failure(error))
+                self.finishSync { completion(.failure(error)) }
```

Comment:
We're calling `completion` twice. That might be confusing for someone who use this method that the `completion` handler is called twice.

Reply:
Omg, sure, I left it here by accident. Thanks for finding that! 🕵️‍♀️